pkg/labels/selector.go

```diff
@@ -216,86 +221,484 @@ func NewRequirement(key string, op Operator, vals util.StringSet) (*Requirement,
+	IN
+	NEQUAL
+	NOTIN
+	OPAR
+	OR
+)
+
+// string2token contains the mapping between lexer Token and token literal
+// (except IDENTIFIER, EOS and ERROR since it makes no sense)
+var string2token = map[string]Token{
+	")":     CPAR,
+	",":     COMMA,
+	"==":    EEQUAL,
+	"=":     EQUAL,
+	"in":    IN,
+	"!=":    NEQUAL,
+	"notin": NOTIN,
+	"(":     OPAR,
+}
+
+// The item produced by the lexer. It contains the Token and the literal.
+type ScannedItem struct {
+	tok     Token
+	literal string
+}
+
+// isWhitespace returns true if the rune is a space, tab, or newline.
+func isWhitespace(ch byte) bool {
+	return ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n'
+}
+
+// isSpecialSymbol detect if the character ch can be an operator
+func isSpecialSymbol(ch byte) bool {
+	switch ch {
+	case '=', '!', '(', ')', ',':
+		return true
+	}
+	return false
+}
+
+// Lexer struct
+type Lexer struct {
+	// s stores the string to be lexed
+	s string
+	// pos is the position currently lexed
+	pos int
+}
+
+// read return the character currently lexed
+// increment the position and check the buffer overflow
+func (l *Lexer) read() (b byte) {
+	b = 0
+	if l.pos < len(l.s) {
+		b = l.s[l.pos]
+		l.pos++
+	}
+	return b
+}
+
+// no return simply unread
+func (l *Lexer) unread() {
+	l.pos--
+}
+
+// func return a literal token (for example IN) and or an identifier.
+func (l *Lexer) scanIdOrKeyword() (tok Token, lit string) {
+	var buffer []byte
+	for {
+		if ch := l.read(); ch == 0 { // end of string found
+			break
+		} else if isSpecialSymbol(ch) || isWhitespace(ch) {
+			l.unread() // stop scanning and unread
+			break
 		} else {
-			reqs = append(reqs, str)
+			buffer = append(buffer, ch)
+		}
+	}
+	s := string(buffer)
+	if val, ok := string2token[s]; ok { // is a literal token?
+		return val, s
+	}
+	return IDENTIFIER, s // otherwise is an identifier
+}
+
+// scan string starting with specail symbol. At the moment this special symbols
+// identify not literal operators
+func (l *Lexer) scanSpecialSymbol() (Token, string) {
+	lastScannedItem := ScannedItem{}
+	var buffer []byte
+	for {
+		if ch := l.read(); ch == 0 {
+			break
+		} else if isSpecialSymbol(ch) {
+			buffer = append(buffer, ch)
+			if token, ok := string2token[string(buffer)]; ok {
+				lastScannedItem = ScannedItem{tok: token, literal: string(buffer)}
+			} else if lastScannedItem.tok != 0 {
+				l.unread()
+				break
+			}
+		} else { // in any other cases (identifer or whitespace) stop
+			l.unread()
+			break
+		}
+	}
+	if lastScannedItem.tok == 0 {
+		return ERROR, fmt.Sprintf("error expected keyword found '%s'", buffer)
+	} else {
+		return lastScannedItem.tok, lastScannedItem.literal
+	}
+}
+
+// func Lex return Token and the literal (meaningfull only in case of IDENTIFIER)
+func (l *Lexer) Lex() (tok Token, lit string) {
+	ch := l.read()
+	for { // consume spaces until no more spaces
+		if !isWhitespace(ch) {
+			break
+		}
+		ch = l.read()
+	}
+	if ch == 0 { // end of the string?
```

Comment:
This is a nit (just a style thing for future pulls): go encourages using a switch whenever you have this sort of pattern - it ends up reading better.  Basically go hates "else if" and wants switch.  If you ever are in here again later it's a minor thing to clean up if you can.